### PR TITLE
[new release] elpi (1.7.0)

### PR DESCRIPTION
**Pull request by gares**
CHANGES:

- Parser:
  - Tolerate trailing `,` in lists, eg `[1,2,3,]` is now parsed as `[1,2,3]`.
  - Error if the input of `Parse.goal_from_string` contains extra tokens
  - Binary conjunction `&` is now turned on the fly into the nary one `,`.

- Bugfix:
  - Nasty bug in pruning during higher order unification, see LPCIC/elpi#36.
  - `Discard` is now considered a stack term, and is turned into an
    unification variable on heapification.
  - `API.RawData.look` now expands `UVar` correctly

- Stdlib:
  - `set` and `map` for arbitrary terms equipped with an order relation.
    Based on the code of OCaml's `Map` and `Set` library.
  - New API `map.remove` for maps on builtin data.

- FFI:
  - New `ContextualConversion` module and `ctx_readback` type. In an FFI call
    one can specify a readback for the hypothetical context that is run once
    and its output is give to the ML code instead of the "raw" constraints and
    hyp list.

- API:
  - Commodity functions in `Elpi.Builtin` to export as built-in
    OCaml's `Set.S` and `Map.S` interfaces (the output of `Set.Make`
    and `Map.Make`). All data is limited to be a closed term.
  - `Constants.andc2` was removed
  - `FlexibleData.Elpi.make` takes no `~lvl` argument (it is always `0`)
  - `RawData.view` no more contains `Discard` since it is not an heap term

- Misc:
  - Pretty printer for unification variable made re-entrant w.r.t. calls to the
    CHR engine (used to lose the mapping between heap cells and names)
  - Pretty printer for solution abstracted over a context (part of the solution). In this
    way the result can be printed correctly even if the runtime has been destroyed.
  - Default paths for finding `.elpi` files fixed after the switch to dune
  - A few more tests regarding unification, data structures and performance

### CODE REVIEW

**Comment by gares** [comment]
There is an issue about `/usr/bin/time` on Alpine linux running the tests, apparently that binary does not accept the options I pass to it.

File: packages/elpi/elpi.1.7.0/opam [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+depends: [
+  "ocaml" {>= "4.04.0"}
+  "camlp5"
+  "ppx_tools_versioned"
+  "ppx_deriving"
+  "ocaml-migrate-parsetree"
+  "re" {>= "1.7.2"}
+  "ANSITerminal" {with-test}
+  "cmdliner" {with-test}
+  "dune"
```
Comment (kit-ty-kate) [inline packages/elpi/elpi.1.7.0/opam]:
```suggestion
  "dune" {>= "1.6"}
```
Your `dune-project` file says you require at least dune 1.6

**Comment by kit-ty-kate** [comment]
> There is an issue about `/usr/bin/time` on Alpine linux running the tests, apparently that binary does not accept the options I pass to it.

Yes, the time binary on Alpine is not the GNU time but the one from Busybox. We can ignore this error or if you prefer you can disable the tests on Alpine or fix it in the source. It's no big deal.

**Comment by gares** [comment]
Please ignore

**Comment by kit-ty-kate** [comment]
Thanks

**Pull request merged**